src/messaging/ExchangeMgr.cpp

```diff
@@ -288,6 +288,12 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
             }
         }
 
+        if (GetDispatchForDelegate(delegate).IsEncryptionRequired() != session->IsEncrypted())
+        {
+            ChipLogError(ExchangeManager, "OnMessageReceived failed, err = %s", ErrorStr(CHIP_ERROR_NO_UNSOLICITED_MESSAGE_HANDLER));
```

Comment:
That's not the error type we used to have here.  Why is this the right type?

Also, this check will _always_ fail if `delegate` is null, right?  As long as we are moving this around, can we skip this check for the `!delegate` case where we're just trying to send a standalone ack?  Or will something go terribly wrong if the session is not in fact encrypted and we try to send an ack via ApplicationExchangeDispatch?